feat: Popular Topics - Permissions (#58)
DisnakeDev/guide: guide/docs/popular-topics/permissions.mdx

## Creating permission objects

`Permissions` can be constructed in two ways: using manual instantiation with granted/revoked permissions specified via
keyword arguments, or shorthand methods that create permission objects with corresponding permissions granted.

:::note
There's actually a third option: specifying the raw value, but it's usage is generally discouraged as it's hard
to understand at a glance what the result will be. Anyway, here's how you do it:

```python
permissions = disnake.Permissions(64)  # Grants the "Add Reactions" permission.

assert permissions.value == 64
```

:::

### Manual creation

Manual creation is usually preferred over shorthand methods due to the flexibility it provides while also still being
as intuitive as possible, and, as such, is also what you'll usually encounter in other examples and documentation.

```python
permissions = disnake.Permissions(
    # Each permission is specified using a keyword argument,
    # where value of `True` means that it is granted, while

>     # `False` revokes the permission.

Author: 9bfce23

>     # `False` revokes the permission.

Author: Missed some references in the above commit, so here's a follow-up: 4a664b3.

for you via a range of methods defined on the <DocsLink reference="disnake.Permissions" /> class.

```python
# All "Advanced" permissions from the official UI granted.
advanced_permissions = disnake.Permissions.advanced()
# ALL permissions granted, including advanced ones.
all_permissions = disnake.Permissions.all()
```

To not duplicate what's already said in <DocsLink reference="disnake.Permissions">docs</DocsLink>, we'll only
list the available methods here:

-   <DocsLink reference="disnake.Permissions.advanced" />
-   <DocsLink reference="disnake.Permissions.all" />
-   <DocsLink reference="disnake.Permissions.all_channel" />
-   <DocsLink reference="disnake.Permissions.events" />
-   <DocsLink reference="disnake.Permissions.general" />
-   <DocsLink reference="disnake.Permissions.membership" />
-   <DocsLink reference="disnake.Permissions.none" />
-   <DocsLink reference="disnake.Permissions.private_channel" />
-   <DocsLink reference="disnake.Permissions.stage" />
-   <DocsLink reference="disnake.Permissions.stage_moderator" />
-   <DocsLink reference="disnake.Permissions.text" />
-   <DocsLink reference="disnake.Permissions.voice" />

> ## Permission Overwrites

Author: 28bca10

<!-- TODO: Overwrites actually consist of a single dict of bools, but users shouldn't really care about that as it

> is an implementation detail (i suppose?). -->

Author: 761c3e921aafc580717f564064fef1e85a9aeb43

<!-- NOTE: The #untouched-permissions anchor is placed a bit above the paragraph to account for website header offset

> This likely won't work for all possible screen/window sizes, but it's better than nothing. -->

Author: The thing is, it's not a header. The `#untouched-permissions` is not even a link - it's just a `<span>` with an `id=untouched-permissions`. Creating a header for that paragraph is not viable at all, but if it's possible without any tricks i'd like to keep the link.

> This likely won't work for all possible screen/window sizes, but it's better than nothing. -->

Author: Never mind, later update will remove the need for the link. Sorry

> This likely won't work for all possible screen/window sizes, but it's better than nothing. -->

Author: 7717fbb

A <DocsLink reference="disnake.PermissionOverwrite" /> is an override for server-wide permissions (i.e., those granted/revoked
by roles). Permission overwrites are assigned per-channel per-user/role and are made up of two <span id="untouched-permissions">`Permissions`</span>
objects: the one containing granted permissions, and the one contaning revoked permissions, collectively called a "pair".

If some permission is not granted neither revoked, then it is considered "untouched" and is not overriden. The value of an

> untouched permission is `None`.

Author: 7717fbb

assert overwrite.pair()[0] == disnake.Permissions(add_reactions=True)
assert overwrite.pair()[1] == disnake.Permissions(view_audit_log=False)
```

You can update overwrites in-place by simply assigning values to attributes or by using
the <DocsLink reference="disnake.PermissionOverwrite.update" /> method.

```python
overwrite.view_audit_log = True

assert overwrite.view_audit_log

overwrite.update(view_audit_log=False)

assert not overwrite.view_audit_log
```

If you need to create a new overwrite object based on an existing one, you can do the following.

```python
pair = overwrite.pair()
new_overwrite = disnake.PermissionOverwrite.from_pair(pair)

new_overwrite.add_reactions = False

assert not new_overwrite.add_reactions
```

## Using permissions

Creating a "chat moderator" role:

```python
permissions = disnake.Permissions(
    manage_messages=True,
    moderate_members=True,
)

chatmod_role = await guild.create_role(
    name="Chat Moderator",
    permissions=permissions,
    hoist=True,
)
```

Overriding chat moderator's permissions in the `#only-mods` <DocsLink reference="disnake.TextChannel"/>:

```python
# We don't want to allow chat mods to delete each other's messages.

> overwrite = disnake.PermissionOverwrite(manage_messages=False)

Author: 4fe392b

    await inter.send("Sure you can!" if inter.permissions.add_reactions else "Sadly no :(")
```

## Querying permissions

There are various ways to query permission information, and they also depend on the context you're
in. This section tries to list all possible methods for getting permission info, grouped by context
they can be used in.

### Context-agnostic

If you have an instance of any class implementing <DocsLink reference="disnake.abc.GuildChannel" />,
then the easiest way to query permissions for a member/role is
through <DocsLink reference="disnake.abc.GuildChannel.permissions_for" />.
If you need a bit more lower level control, you can also get raw overwrite objects
using <DocsLink reference="disnake.abc.GuildChannel.overwrites_for" />
or <DocsLink reference="disnake.abc.GuildChannel.overwrites" />.

### In prefix commands

There are no direct methods on <DocsLink reference="disnake.ext.commands.Context" /> for querying permission
information, however, you can still access <DocsLink reference="disnake.ext.commands.Context.channel" /> and
call above-mentioned methods on it, though be sure to check that it's not <DocsLink reference="disnake.PartialMessageable" />

> as it doesn't have any permissions-related methods on it either.

Author: Technically it is (according to the union type of Context.channel)

> as it doesn't have any permissions-related methods on it either.

Reviewer: Ah, right. I was only looking at the type hint, but it's indeed mentioned in the documentation

defined on <DocsLink reference="disnake.Interaction" />: <DocsLink reference="disnake.Interaction.permissions" />
(for getting interaction author's permissions) and <DocsLink reference="disnake.Interaction.app_permissions" />
(for getting bot's permissions in interaction's channel).

## How permissions are calculated

Permissions can be a tough thing to understand at first, but understanding how permissions are
applied can help you <!-- feel the pain of us mere mortals re-implementing discord client's obscure behaviour -->
better grasp this concept.

In case we're calculating permissions for a member, the strategy is as follows:

-   If member is the guild owner, all permissions are granted - no questions asked. Otherwise..
-   First we apply the `@everyone` role's permissions.
-   Then, for each role the member has we apply it's permissions too.
-   -   If at this point the member has the "Administrator" permission, all others are granted automatically
        (the member is a guild-wide administrator, and, as such, is functionally equivalent to the owner except
        they can't delete the guild).

> -   After that, if we're in a channel, the following is additionally done.

Reviewer: Doesn't this always happen in a channel? I might be missing something here, though.

Author: While disnake does not provide any abstractions for that, one can in theory calculate permissions on a guild-level (for whatever reason). Since this section talks about permission calculation as a whole and not about concrete means to do it in disnake, i personally think this has a place to be.

Reviewer: Alright, fair enough!

-   -   We apply it's overwrites in the same order: `@everyone` overwrites,
        role overwrites and finally member overwrites.
-   -   Then [channel-specific implicit permissions](#channel-specific-implicit-permissions) are handled.

> -   In the end, we take into account [guild-wide implicit permissions](#guild-wide-implicit-permissions).

Reviewer: this should probably be an enumeration

Author: a15b78b18fc6f63696fa98d4ed842b5095386688

In case of calculating permissions for a role, the strategy is the same as if we were calculating permissions
for a member with only that role (and `@everyone`), except that member overwrites are
not taken into account.

:::note
Curious about the implementation? You can see it [here](https://git.disnake.dev/disnake/blob/master/disnake/abc.py#L651-L808).

> :::

Reviewer: This link will change over time, maybe point to https://discord.com/developers/docs/topics/permissions#permission-overwrites instead? :thinking:

Author: The note is about how disnake implements permission calculation, not how discord tells to do it. I also now remembered that we can pin link to a specific revision, which is done in 26020fb24bc0ac78072e1f577ca445723d6c41a0. Please unresolve conversation if this doesn't fit you.

Reviewer: Linking to a specific ref seems good, sure.
Another approach would be instructing readers to use the `[source]` button [here](https://docs.disnake.dev/en/stable/api/abc.html#disnake.abc.GuildChannel.permissions_for), but that's sort of unnecessary here since the implementation really shouldn't be changing a whole lot.

### Implicit permissions

In Discord, there are some permissions that, when revoked, basically make it impossible to utilize other
granted permissions. For example, take a user with two permissions: "View Channel" and
"Send Messages". If at some point user loses the "View Channel" permission, it wouldn't matter whether
they can send messages in that channel: they just can't see the channel in the UI. disnake's permission system
attempts to handle such cases via "implicit permissions" - permissions that, when granted or, more commonly,
revoked, automatically grant (revoke) other permissions as well. The currently handled cases are separated
into two parts: guild-wide rules and channel-specific rules, and are described below.

#### Guild-wide implicit permissions

> Guild-wide implicit permissions currently consist of a single rule.

Reviewer: this line seems unnecessary

Author: 94f9216d037592b33d3c38f2adf516f62bb5736f

> 1. Timed out members are spectators.

Reviewer: Consider mentioning that this part of the permission handling can be turned off using `.permissions_for(..., ignore_timeout=True)`

Author: Isn't this a legacy compatibility parameter? I don't think it's usage is a good idea

Reviewer: @onerandomusername should know more about this, I'm not entirely sure if we planned to deprecate it.

Reviewer: On second thought, you're right, this is fine to leave as-is. Even if we don't end up removing the parameter in later versions, it doesn't have to be mentioned here

This rule autmatically revokes all permissions for timed out members _except_ "View Channel" and
"Read Message History", which are left [untouched](#untouched-permissions).

#### Channel-specific implicit permissions

Channel-specific implicit permissions currently include two rules.

1. If you can't send a message in a channel, then you can't sent _any_ message in it.

This rule automatically revokes "Send TTS Messages", "Send Voice Messages", "Mention Everyone", "Embed Links"
and "Attach Files" permissions for members that do not have the "Send Messages" permission.

> 2. If you can't view a channel, they you can't do anything in it.

Reviewer: Something similar also happens with the `Connect` permission in voice/stage channels

Author: b84e034

This rule automatically revokes "All Channel" permissions, as specified by <DocsLink reference="disnake.Permissions.all_channel" />,

> if the member doesn't have the "View Channel" permission.

Reviewer: I'm not sure if we want to document all of these - Discord doesn't either (it is, and I quote, ["based on logical conclusions"](https://discord.com/developers/docs/topics/permissions#implicit-permissions)).
If anything, mentioning that this is not an exhaustive list would make sense, as there are a bunch of fun edge cases

Author: I rephrased this a bit so that these rules only act as examples, and by no means are an exhaustive list of all handled/possible cases where implicit permissions come into action. 537115d0ef62104ffc522159944da587cc554de6

```python
base = disnake.Permissions(add_reactions=True, manage_guild=True)
same = disnake.Permissions(add_reactions=True, manage_guild=True)
superset = disnake.Permissions(add_reactions=True, manage_guild=True, send_messages=True)
subset = disnake.Permissions(add_reactions=True)
different = disnake.Permissions(use_voice_activation=True)
```

### Basic

These are operations that usually require no explanation. In any way, comments will guide you through.

```python
# Check that two permission object have same permissions set to the same value.
assert base == same
# Check that one (base) permission object has at least all of the granted permissions
# of the second one granted too.
assert base >= same
# Check that one (base) permission object has all of the granted permissions
# of the second one granted too, plus at least one extra.
assert base > subset
# Check that one (base) permission object has at maximum all of the granted permissions
# of the second one granted.
assert base <= same
# Check that one (base) permission object has at maximum of the granted permissions
# of the second one granted, minus one or more.
assert base < superset

> # Check that one (base) permission object has at least one permission granted/revoked
> # that is revoked/granted in the second one.

Reviewer: ```suggestion
# Check that one (base) permission object has at least one permission allowed
# that is denied in the second one (and the other way around).
```

assert base != different
```

### Advanced

If you're familar with bit operators in Python, then you'll feel right at home. Otherwise, read
the comments to understand what does what.

```python

> # Creates a new Permissions object with all granted permissions from both base and different.

Reviewer: ```suggestion
# Creates a new Permissions object with all granted permissions from either one of them.
```

assert subset | different == disnake.Permissions(add_reactions=True, use_voice_activation=True)

> # Creates a new Permissions object with only permissions granted on both base and different.

Reviewer: ```suggestion
# Creates a new Permissions object with only permissions granted on both.
```

assert base & subset == disnake.Permissions(add_reactions=True)
# Creates a new Permissions object with only permissions granted on one of base or different,
# but not both.
assert base ^ different

# Notably, all of operators above support in-place updates using the `OP=` syntax, like so:

> # `base &= subset`.

Reviewer: This part probably doesn't need to be mentioned here

Author: Did you mean it should be mentioned somewhere else or it doesn't need to be mentioned at all (in this case, why?)?

Reviewer: My bad, I should've been more clear - I meant removing it entirely.
However, thinking about it again, keeping this line is also fine, so disregard my comment

# Creates a new Permissions object with all permissions from subset inverted.
assert not (~subset).add_reactions and (~subset).manage_guild
# Creates a new Permissions object with all permissions except add_reactions granted.
assert not (~disnake.Permissions.add_reactions).add_reactions
# Creates a new Permissions object with all specified permissions granted.
assert disnake.Permissions.add_reactions | disnake.Permissions.manage_guild == base
assert subset | disnake.Permissions.manage_guild == base
```

### Additional

There's also a few additional operations supported that do not involve operators.

```python
# Returns the permissions's hash. This currently returns it's value, but that may change
# in the future.

> assert hash(base) == base.value == 96

Reviewer: I'd also argue it's better to not mention this here, or at least not that it's equal to `.value` which is generally an implementation detail

Author: Reworked in 22e7574.

# Permissions object implements `__iter__()`, and thus can be used in for loops.
for perm, value in base:
    truthy = []

    if value:
        truthy.append(perm)

>     assert truthy == ["add_reactions", "manage_guild"]

Reviewer: imo this might be worth its own section, it's often overlooked and is supported by both `Permissions` and `PermissionOverwrite`

Author: I'm not quite sure how to fill such section (there isn't anything much to talk about or to explain). By the way fixed the example (b84f27e).

Reviewer: Fair point, it would be difficult to move this into a separate section since the functionality is pretty straightforward.
Added another comment here, but overall this looks great: https://github.com/DisnakeDev/guide/pull/58#discussion_r1258096528

This also works with <DocsLink reference="disnake.Thread" />s.
If you need a bit more lower level control, you can also get raw overwrite objects
using <DocsLink reference="disnake.abc.GuildChannel.overwrites_for" />
or <DocsLink reference="disnake.abc.GuildChannel.overwrites" />.

### In prefix commands

There are no direct methods on <DocsLink reference="disnake.ext.commands.Context" /> for querying permission
information, however, you can still access <DocsLink reference="disnake.ext.commands.Context.channel" /> and
call above-mentioned methods on it, though be sure to check that it's not <DocsLink reference="disnake.PartialMessageable" />
as it doesn't have any permissions-related methods on it either.

### In interactions

When working with interactions (i.e., with application commands or components), there are two main properties
defined on <DocsLink reference="disnake.Interaction" />: <DocsLink reference="disnake.Interaction.permissions" />
(for getting the interaction author's permissions in the current channel) and <DocsLink reference="disnake.Interaction.app_permissions" />
(for getting the bot's permissions).

## How permissions are calculated

> Permissions can be a tough thing to understand at first, but understanding how permissions are
> applied can help you <!-- feel the pain of us mere mortals re-implementing discord client's obscure behaviour -->
> better grasp this concept.

Reviewer: ```suggestion
Permissions can be a tough thing to understand at first, but understanding how permissions are
applied can help you better grasp this concept.
```
github's syntax highlighter really doesn't like this for some reason, so we should probably just remove the comment c:
it do be like that though.

In case we're calculating permissions for a member, the strategy is as follows:

1.  If member is the guild owner, all permissions are granted - no questions asked. Otherwise..

2.  First we apply the `@everyone` role's permissions.

> 3.  Then, for each role the member has we apply it's permissions too.
> 
> 4.  If at this point the member has the "Administrator" permission, all others are granted automatically
>     (the member is a guild-wide administrator, and, as such, is functionally equivalent to the owner except
>     they can't delete the guild).
> 
> 5.  After that, if we're in a channel, the following is additionally done.
> 
> 6.  1.  We apply it's overwrites in the same order: `@everyone` overwrites,
>         role overwrites and finally member overwrites.
> 
>     2.  Then [channel-specific implicit permissions](#channel-specific-implicit-permissions) are handled.
> 
> 7.  In the end, we take into account [guild-wide implicit permissions](#guild-wide-implicit-permissions).

Reviewer: ```suggestion
3.  Then, for each of the member's roles, apply its permissions too.

4.  If at this point the member has the "Administrator" permission, all other permissions are granted automatically
    (the member is a guild-wide administrator, and, as such, is functionally equivalent to the owner except
    they can't delete the guild).

5.  After that, if we're in a channel, the following is additionally done:

    1.  We apply its overwrites in the same order: `@everyone` overwrites,
        role overwrites and finally member overwrites.

    2.  Then [channel-specific implicit permissions](#channel-specific-implicit-permissions) are handled.

6.  In the end, we take into account [guild-wide implicit permissions](#guild-wide-implicit-permissions).
```
tiny fixes and merging 5. + 6. into one element

assert base & subset == disnake.Permissions(add_reactions=True)
# Creates a new Permissions object with only permissions allowed on one of base or different,
# but not both.
assert base ^ different

# Notably, all of operators above support in-place updates using the `OP=` syntax, like so:
# `base &= subset`.

# Creates a new Permissions object with all permissions from subset inverted.
assert not (~subset).add_reactions and (~subset).manage_guild
# Creates a new Permissions object with all permissions except add_reactions allowed.
assert not (~disnake.Permissions.add_reactions).add_reactions
# Creates a new Permissions object with all specified permissions allowed.
assert disnake.Permissions.add_reactions | disnake.Permissions.manage_guild == base
assert subset | disnake.Permissions.manage_guild == base
```

### Additional

There's also a few additional operations supported that do not involve operators.

```python
# Returns the permissions's hash. Allows permission objects to be used as dictionary keys.
hash(base)

# Permissions and PermissionOverwrite both implement `__iter__()`, and thus can be used in for loops.

> truthy = []
> 
> for perm, value in base:
>     if value:
>         truthy.append(perm)
> 
> assert truthy == ["add_reactions", "manage_guild"]

Reviewer: ```suggestion
allowed = []

for name, value in base:
    if value:
        allowed.append(name)

assert allowed == ["add_reactions", "manage_guild"]
```
